### PR TITLE
Feature: Support for Ambiq Apollo 3

### PR DESCRIPTION
**Pull request by sidprice**
This PR is a partial implementation of support for the Ambiq Apollo 3 Blue MCU. Due to the lack of documentation for flash erase and program, those features are not implemented.

Note: The MCU has a set of functions in a built-in ROM Bootloader that implement the required functions, however, there is no source code available or documentation describing how they should be used. In addition, implementing access to these functions for BMDA is complex.

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (`make PROBE_HOST=native`)
* [x] It builds as BMDA (`make PROBE_HOST=hosted`)
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

### CODE REVIEW

File: src/target/apollo3.c [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+ */
+
+#include "general.h"
+#include "target.h"
+#include "target_internal.h"
+
+static bool apollo_3_flash_erase(target_flash_s *flash, target_addr_t addr, size_t len);
+static bool apollo_3_flash_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t len);
+
+#define FLASH_BASE_ADDRESS 0x00000000U
```
Comment (dragonmux) [inline src/target/apollo3.c:28]: Ideally all defines in this file would be prefixed with `APOLLO_3_` though this is an optional change

File: src/target/apollo3.c [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+#define CHIPPN_QUALIFIED_MASK         0x00000001U
+#define CHIPPN_QUALIFIED_BIT_POSITION 0x0U
+
+#define CHIPID0_REGISTER 0x40020004U // Chip ID Register 0
+#define CHIPID1_REGISTER 0x40020008U // Chip ID Register 1
+
+#define CHIPREV_REGISTER 0x4002000cU // Chip Revision Register
+
+/*
```
Comment (dragonmux) [inline src/target/apollo3.c:71]: Please switch the comment style here to:
```c
/*
 * A Multi-line
 * commment
 */

File: src/target/apollo3.c [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+bool apollo_3_probe(target_s *target)
+{
+	uint32_t mcu_vendor_id = target_mem_read32(target, VENDOR_ID_ADDRESS);
+	if (mcu_vendor_id != VENDOR_ID) {
+		DEBUG_INFO("Invalid vendor ID read\n");
+		return false;
+	} else
+		DEBUG_INFO("Read correct vendor ID\n");
+	/*
```
Comment (dragonmux) [inline src/target/apollo3.c:109]: This and the comment below need switching either to proper multi-line style, or collapsing to single-line please.

File: src/target/apollo3.c [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+static bool apollo_3_flash_erase(target_flash_s *flash, target_addr_t addr, size_t len);
+static bool apollo_3_flash_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t len);
+
+#define FLASH_BASE_ADDRESS 0x00000000U
+#define FLASH_SIZE         0x00100000
+#define FLASH_BLOCK_SIZE   0x2000U
+
+#define SRAM_BASE 0x10000000
+#define SRAM_SIZE 0x00060000
```
Comment (dragonmux) [inline src/target/apollo3.c:33]: These are missing their `U` suffixes please.

**Review by dragonmux — CHANGES_REQUESTED** [review]
👋🏼 Sid, there are a few stylistic items that need addressing, but nothing terrible - once they're sorted we'll be happy to merge this as preliminary support.

File: src/target/apollo3.c (hunk 2 above)
Comment (dragonmux) [reply src/target/apollo3.c:71]: It seems this one comment got missed, but with that taken care of this is all good to merge - excellent work to both Sid and litui!

Comment (litui) [reply src/target/apollo3.c:71]: Whoops could have sworn I got all those. Ah well. =). This one I'll leave to you @sidprice.

Comment (dragonmux) [reply src/target/apollo3.c:71]: Agreed to take this as is and do a fixup once merged. Resolving and approving 🙂

**Review by dragonmux — APPROVED** [review]

**Comment by dragonmux** [comment]
LGTM, merging. Thank you both for the contribtuion!

**Pull request merged**